EIPS/eip-draft_transaction_data_opcodes.md

```diff
@@ -0,0 +1,106 @@
+---
+eip: <to be assigned>
+title: Transaction Data Opcodes
+author: Alex Papageorgiou (@alex-ppg)
+discussions-to: https://ethereum-magicians.org/t/eip-draft-transaction-data-opcodes/6017
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-04-16
+---
+
+## Simple Summary
+
+Provide access to original transaction data.
+
+## Abstract
+
+This EIP introduces the following four EVM instructions: `ORIGINDATALOAD`, `ORIGINDATASIZE`, `ORIGINDATACOPY` and `ENTRYPOINT`.
+
+The first three instructions are meant to provide access to the original transaction's `data` payload whereas the last instruction is meant to provide access to the original recipient of the transaction, the `to` address.
+
+## Motivation
+
+It is undeniable that smart contracts are becoming more interconnected than ever. Up until this point, smart contracts have entirely relied on compliant interfaces and introspection to introduce a new step in the call chain of a complex multi-contract interaction. However, this presents a forwards-only approach which limits the types of interactions that can manifest.
+
+The purpose of this EIP is to provide a way via which a contract is able to identify the entry-point of a transaction on the blockchain and deduce what was the original intention of the transaction by applying introspection on the original transaction data itself.
+
+This EIP enables the development of new types of smart contracts as it can open new pathways for [EIP-721](https://eips.ethereum.org/EIPS/eip-721) NFTs and [EIP-20](https://eips.ethereum.org/EIPS/eip-20) tokens to detect which action their transaction is part of, such as detecting a liquidity provision to a decentralized exchange or a loan within a collateralized lending protocol.
+
+As a side-effect, it acts as an enhanced "substitute" to the `ORIGIN` check that the [EIP-3074](https://eips.ethereum.org/EIPS/eip-3074) breaks, enabling a greater level of granularity to be applied to the introspection of a transaction f.e. by validating the `ENTRYPOINT` being equal to the currently executed contract.
+
+## Specification
+
+### ORIGINDATALOAD (`0x47`), ORIGINDATASIZE (`0x48`) and ORIGINDATACOPY (`0x49`)
+
+These instructions are meant to operate identically to their `CALL`-prefixed counterparts with the exception that they instead operate on the original `data` of a transaction instead of the current call's data. As the data is retrieved once again from the execution environment, the costs for the three instructions will be `G_verylow`, `G_base` and `G_base + G_verylow * (number of words copied, rounded up)` respectively.
```

Comment:
When referring to existing opcodes, use their full name and opcode value like `ORIGINDATALOAD (0x47)`.  In this case, please list the CALL opcodes you are referring to along with their values.

Reply:
Thanks, I adopted this change in the latest commit. Kindly let me know if it looks good to you.